txnsync/peer_test.go

```diff
@@ -0,0 +1,871 @@
+		txid := transactions.Txid{byte(i)}
+		a.False(p.recentSentTransactions.contained(txid))
+	}
+
+	p.updateIncomingTransactionGroups(txnGroups)
+
+	for i := 0; i < 10; i++ {
+		txid := transactions.Txid{byte(i)}
+		a.True(p.recentSentTransactions.contained(txid))
+	}
+}
+
+// TestUpdateRequestParams tests updating the request parameters
+func TestUpdateRequestParams(t *testing.T) {
+	a := require.New(t)
+	p := makePeer(nil, true, true)
+	oldModulator := p.requestedTransactionsModulator
+	oldOffset := p.requestedTransactionsOffset
+
+	p.updateRequestParams(oldModulator, oldOffset)
+	a.Equal(p.requestedTransactionsModulator, oldModulator)
+	a.Equal(p.requestedTransactionsOffset, oldOffset)
+
+	p.updateRequestParams(oldModulator+1, oldOffset+1)
+	a.Equal(p.requestedTransactionsModulator, oldModulator+1)
+	a.Equal(p.requestedTransactionsOffset, oldOffset+1)
+
+}
+
+// TestAddIncomingBloomFilter tests adding an incoming bloom filter
+func TestAddIncomingBloomFilter(t *testing.T) {
+	a := require.New(t)
+	p := makePeer(nil, true, true)
+
+	for i := 0; i < 2*maxIncomingBloomFilterHistory; i++ {
+		bf := bloomFilter{
+			encodingParams: requestParams{
+				_struct:   struct{}{},
+				Offset:    byte(i),
+				Modulator: 0,
+			},
+			filter:             nil,
+			containedTxnsRange: transactionsRange{},
+			encoded:            nil,
+			filterType:         0,
+		}
+		p.addIncomingBloomFilter(basics.Round(i), bf, basics.Round(i))
+	}
+
+	a.Equal(len(p.recentIncomingBloomFilters), 2)
+
+	for i := 0; i < 2*maxIncomingBloomFilterHistory; i++ {
+		bf := bloomFilter{
+			encodingParams: requestParams{
+				_struct:   struct{}{},
+				Offset:    byte(i),
+				Modulator: 0,
+			},
+			filter:             nil,
+			containedTxnsRange: transactionsRange{},
+			encoded:            nil,
+			filterType:         0,
+		}
+		p.addIncomingBloomFilter(basics.Round(i), bf, 0)
+	}
+
+	a.Equal(len(p.recentIncomingBloomFilters), maxIncomingBloomFilterHistory)
+
+}
+
+// TestSelectPendingTransactions tests selectPendingTransactions
+func TestSelectPendingTransactions(t *testing.T) {
+
+	type args struct {
+		pendingTransactions []transactions.SignedTxGroup
+		sendWindow          time.Duration
+		round               basics.Round
+		bloomFilterSize     int
+	}
+
+	type results struct {
+		selectedTxns           []transactions.SignedTxGroup
+		selectedTxnIDs         []transactions.Txid
+		partialTranscationsSet bool
+	}
+
+	tests := []struct {
+		name   string
+		fxn    func(p *Peer)
+		arg    args
+		result results
+	}{
+		{"Case 1", func(p *Peer) { p.lastRound = 98 }, args{nil, time.Millisecond, 100, 0}, results{nil, nil, false}},
+		{"Case 2", func(p *Peer) { p.lastRound = 101; p.requestedTransactionsModulator = 0 }, args{nil, time.Millisecond, 100, 0}, results{nil, nil, false}},
+		{"Case 3", func(p *Peer) { p.lastRound = 200; p.messageSeriesPendingTransactions = nil }, args{[]transactions.SignedTxGroup{}, time.Millisecond, 100, 0}, results{nil, nil, false}},
```

Comment:
Is there any logic here testing using the modulator / max size of the returned list of txgroups?

Reply:
None specifically here but I tested SignedTxGroup functionality in this file.  Are there limits/specific tests you would recommend?

Reply:
I suggest doing something simple ie have a limit of 2 transactions and a modulator of 2 with 6 total txns. This way, you can check if something like only txns 1 and 3 were selected.